tools/sklearn/utils.py

```diff
@@ -222,11 +254,112 @@ def get_search_params(params_builder):
         lst = search_p.split(":")
         assert (len(lst) == 2), "Error, make sure there is one and only one colon in search parameter input."
         literal = lst[1].strip()
-        ev = safe_eval(literal)
-        if param_type == "final_estimator_p":
-            search_params["estimator__" + lst[0].strip()] = ev
+        param_name = lst[0].strip()
+        if param_name:
+            if param_name.lower() == 'n_jobs':
+                sys.exit("Parameter `%s` is invalid for search." %param_name)
+            elif not param_name.endswith('-'):
+                ev = safe_eval(literal)
+                if param_type == "final_estimator_p":
+                    search_params["estimator__" + param_name] = ev
+                else:
+                    search_params["preprocessing_" + param_type[5:6] + "__" + param_name] = ev
+            else:
+                # only for estimator eval, add `-` to the end of param
+                #TODO maybe add regular express check
+                ev = safe_eval_es(literal)
+                for obj in ev:
+                    if 'n_jobs' in obj.get_params():
+                        obj.set_params( n_jobs=N_JOBS )
+                if param_type == "final_estimator_p":
+                    search_params["estimator__" + param_name[:-1]] = ev
+                else:
+                    search_params["preprocessing_" + param_type[5:6] + "__" + param_name[:-1]] = ev
+        elif param_type != "final_estimator_p":
+            #TODO regular express check ?
+            ev = safe_eval_es(literal)
+            preprocessors = [preprocessing.StandardScaler(), preprocessing.Binarizer(), preprocessing.Imputer(),
```

Comment:
I find it more readable if you create a separate list for all the sub-groups to not do the slicing later preprocessors[0:8] and so on..

Reply:
Probably you're right. My solution may not be the best, or even one of those. I have no better idea on how to handle the hot numbers, and the mix at current moment. If you have one, a PR or commit is welcome